From fc4e9df4e81bd9dee68112122348df52a046fe1b Mon Sep 17 00:00:00 2001
From: Zhiwei Yang <quic_zhiwyang@quicinc.com>
Date: Wed, 7 Dec 2022 14:42:26 +0800
Subject: [PATCH] qcacld-3.0: Support 4 virtual interfaces

Support 4 virtual interfaces,include 1STA and 3SAP
in the DBS/SCC mode.

Signed-off-by: zhiwyang <quic_zhiwyang@quicinc.com>
---
 .../policy_mgr/inc/wlan_policy_mgr_api.h           |  96 ++++++
 .../policy_mgr/inc/wlan_policy_mgr_public_struct.h |  23 ++
 .../policy_mgr/src/wlan_policy_mgr_action.c        |  15 +-
 .../policy_mgr/src/wlan_policy_mgr_core.c          | 324 +++++++++++++--------
 .../policy_mgr/src/wlan_policy_mgr_get_set_utils.c |  65 ++++-
 .../policy_mgr/src/wlan_policy_mgr_i.h             |   3 +
 .../policy_mgr/src/wlan_policy_mgr_init_deinit.c   |   3 +
 .../policy_mgr/src/wlan_policy_mgr_pcl.c           | 153 +++++++++-
 .../src/wlan_policy_mgr_tables_2x2_dbs_i.h         |  37 ++-
 core/hdd/src/wlan_hdd_conc_ut.c                    |  40 +--
 core/hdd/src/wlan_hdd_hostapd.c                    |   4 +-
 11 files changed, 601 insertions(+), 162 deletions(-)

diff --git a/components/cmn_services/policy_mgr/inc/wlan_policy_mgr_api.h b/components/cmn_services/policy_mgr/inc/wlan_policy_mgr_api.h
index 234ada1..9e51b06 100644
--- a/components/cmn_services/policy_mgr/inc/wlan_policy_mgr_api.h
+++ b/components/cmn_services/policy_mgr/inc/wlan_policy_mgr_api.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2021 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -49,6 +50,11 @@ typedef const enum policy_mgr_conc_next_action
 	policy_mgr_next_action_three_connection_table_type
 	[PM_MAX_TWO_CONNECTION_MODE][POLICY_MGR_MAX_BAND];
 
+typedef const enum policy_mgr_conc_next_action
+	policy_mgr_next_action_four_connection_table_type
+	[PM_MAX_THREE_CONNECTION_MODE][POLICY_MGR_MAX_BAND];
+
+
 #define PM_FW_MODE_STA_STA_BIT_POS       0
 #define PM_FW_MODE_STA_P2P_BIT_POS       1
 
@@ -115,6 +121,52 @@ enum PM_AP_DFS_MASTER_MODE {
 	PM_STA_SAP_ON_DFS_MASTER_MODE_FLEX,
 };
 
+static inline const char *device_mode_to_string(uint8_t idx)
+{
+	switch (idx) {
+	CASE_RETURN_STRING(PM_STA_MODE);
+	CASE_RETURN_STRING(PM_SAP_MODE);
+	CASE_RETURN_STRING(PM_P2P_CLIENT_MODE);
+	CASE_RETURN_STRING(PM_P2P_GO_MODE);
+	default:
+		return "none";
+	}
+}
+
+static inline const char *pcl_type_to_string(uint8_t idx)
+{
+	switch (idx) {
+	CASE_RETURN_STRING(PM_NONE);
+	CASE_RETURN_STRING(PM_24G);
+	CASE_RETURN_STRING(PM_5G);
+	CASE_RETURN_STRING(PM_SCC_CH);
+	CASE_RETURN_STRING(PM_MCC_CH);
+	CASE_RETURN_STRING(PM_SCC_CH_24G);
+	CASE_RETURN_STRING(PM_SCC_CH_5G);
+	CASE_RETURN_STRING(PM_24G_SCC_CH);
+	CASE_RETURN_STRING(PM_5G_SCC_CH);
+	CASE_RETURN_STRING(PM_SCC_ON_5_SCC_ON_24_24G);
+	CASE_RETURN_STRING(PM_SCC_ON_5_SCC_ON_24_5G);
+	CASE_RETURN_STRING(PM_SCC_ON_24_SCC_ON_5_24G);
+	CASE_RETURN_STRING(PM_SCC_ON_24_SCC_ON_5_5G);
+	CASE_RETURN_STRING(PM_SCC_ON_5_SCC_ON_24);
+	CASE_RETURN_STRING(PM_SCC_ON_24_SCC_ON_5);
+	CASE_RETURN_STRING(PM_MCC_CH_24G);
+	CASE_RETURN_STRING(PM_MCC_CH_5G);
+	CASE_RETURN_STRING(PM_24G_MCC_CH);
+	CASE_RETURN_STRING(PM_5G_MCC_CH);
+	CASE_RETURN_STRING(PM_SBS_CH_5G);
+	CASE_RETURN_STRING(PM_24G_SCC_CH_SBS_CH);
+	CASE_RETURN_STRING(PM_24G_SCC_CH_SBS_CH_5G);
+	CASE_RETURN_STRING(PM_24G_SBS_CH_MCC_CH);
+	CASE_RETURN_STRING(PM_SCC_ON_2G);
+	CASE_RETURN_STRING(PM_SCC_ON_5G);
+	CASE_RETURN_STRING(PM_5G_24G);
+	default:
+		return "Unknown";
+	}
+}
+
 /**
  * policy_mgr_get_allow_mcc_go_diff_bi() - to get information on whether GO
  *						can have diff BI than STA in MCC
@@ -3661,4 +3713,48 @@ bool policy_mgr_is_sta_mon_concurrency(struct wlan_objmgr_psoc *psoc);
  *
  */
 QDF_STATUS policy_mgr_check_mon_concurrency(struct wlan_objmgr_psoc *psoc);
+
+/**
+ * policy_mgr_is_3rd_conn_on_same_band_allowed() - Check the third connection
+ * on same band allowed or not
+ * list for third connection
+ * @psoc: PSOC object information
+ * @mode: Device mode
+ * @ch_freq: 3rd channel frequency
+ *
+ * This function checks whether to allow third connection on same band or not
+ * based on pcl table
+ *
+ * Return: TRUE/FALSE
+ */
+bool policy_mgr_is_3rd_conn_on_same_band_allowed(struct wlan_objmgr_psoc *psoc,
+						 enum policy_mgr_con_mode mode,
+						 qdf_freq_t ch_freq);
+
+/**
+ * policy_mgr_allow_4th_new_freq() - Function to check whether 4th freq can
+ * be allowed wthout leading to 3 home freq on same mac
+ * @psoc: Pointer to Psoc
+ * @freq1: Frequency 1
+ * @freq2: Frequency 2
+ * @freq3: Frequency 3
+ * @new_ch_freq: freq to check with reference to freq1 freq2 and freq3
+ *
+ * Return:True if all 4 freq can be allowed without causing 3 home frequency
+ * on same mac
+ */
+#ifdef FEATURE_FOURTH_CONNECTION
+bool
+policy_mgr_allow_4th_new_freq(struct wlan_objmgr_psoc *psoc,
+			      qdf_freq_t freq1, qdf_freq_t freq2,
+			      qdf_freq_t freq3, qdf_freq_t new_ch_freq);
+#else
+static inline bool
+policy_mgr_allow_4th_new_freq(struct wlan_objmgr_psoc *psoc,
+			      qdf_freq_t freq1, qdf_freq_t freq2,
+			      qdf_freq_t freq3, qdf_freq_t new_ch_freq)
+{
+	return false;
+}
+#endif
 #endif /* __WLAN_POLICY_MGR_API_H */
diff --git a/components/cmn_services/policy_mgr/inc/wlan_policy_mgr_public_struct.h b/components/cmn_services/policy_mgr/inc/wlan_policy_mgr_public_struct.h
index f272a5c..0120df6 100644
--- a/components/cmn_services/policy_mgr/inc/wlan_policy_mgr_public_struct.h
+++ b/components/cmn_services/policy_mgr/inc/wlan_policy_mgr_public_struct.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2020 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -171,6 +172,8 @@ enum policy_mgr_pcl_group_id {
  * @POLICY_MGR_PCL_ORDER_NONE: no order
  * @POLICY_MGR_PCL_ORDER_24G_THEN_5G: 2.4 Ghz channel followed by 5 Ghz channel
  * @POLICY_MGR_PCL_ORDER_5G_THEN_2G: 5 Ghz channel followed by 2.4 Ghz channel
+ * @POLICY_MGR_PCL_ORDER_24G_ONLY: 2.4 Ghz channel only
+ * @POLICY_MGR_PCL_ORDER_5G_ONLY: 5 Ghz channel only
  *
  * Order in which the PCL is requested
  */
@@ -178,6 +181,8 @@ enum policy_mgr_pcl_channel_order {
 	POLICY_MGR_PCL_ORDER_NONE,
 	POLICY_MGR_PCL_ORDER_24G_THEN_5G,
 	POLICY_MGR_PCL_ORDER_5G_THEN_2G,
+	POLICY_MGR_PCL_ORDER_24G_ONLY,
+	POLICY_MGR_PCL_ORDER_5G_ONLY,
 };
 
 /**
@@ -318,6 +323,9 @@ enum policy_mgr_mac_use {
  * @PM_24G_SCC_CH_SBS_CH_5G: 2.4 Ghz channels, SCC channel,
  *      SBS channels & rest of the 5G channels
  * @PM_24G_SBS_CH_MCC_CH: 2.4 Ghz channels, SBS channels & MCC channels
+ * @PM_SCC_ON_2G: Only SCC channel on 2.4 Ghz
+ * @PM_SCC_ON_5G: Only SCC channel on 5 Ghz
+ * @PM_5G_24G: 5G channels & 2.4 Ghz channels
  * @PM_MAX_PCL_TYPE: Max place holder
  *
  * These are generic IDs that identify the various roles
@@ -348,6 +356,9 @@ enum policy_mgr_pcl_type {
 	PM_24G_SCC_CH_SBS_CH,
 	PM_24G_SCC_CH_SBS_CH_5G,
 	PM_24G_SBS_CH_MCC_CH,
+	PM_SCC_ON_2G,
+	PM_SCC_ON_5G,
+	PM_5G_24G,
 
 	PM_MAX_PCL_TYPE
 };
@@ -802,6 +813,12 @@ enum policy_mgr_two_connection_mode {
  * 5ghz DBS
  * @PM_NDI_NDI_5_NAN_DISC_24_DBS: Both NDI on 5ghz and NAN Disc on 2.4Ghz DBS
  * @PM_NDI_NDI_NAN_DISC_24_SMM: Both NDI, NAN Disc on 2.4ghz SMM
+ * @PM_SAP_SAP_SCC_24_STA_5_DBS: Both SAP on 2.4Ghz and another STA on 5Ghz DBS
+ * @PM_SAP_SAP_SCC_24_STA_24_SCC: Both SAP on 2.4Ghz and another STA on 2.4Ghz SCC
+ * @PM_SAP_SAP_SCC_24_SAP_5_DBS: Both SAP on 2.4Ghz and another SAP on 5Ghz DBS
+ * @PM_SAP_SAP_SCC_5_STA_24_DBS: Both SAP on 5Ghz and another STA on 2.4Ghz DBS
+ * @PM_SAP_SAP_SCC_5_STA_5_SCC: Both SAP on 5Ghz and another STA on 5Ghz SCC
+ * @PM_SAP_SAP_SCC_5_SAP_24_DBS: Both SAP on 5Ghz and another SAP on 2.4Ghz DBS
  */
 enum policy_mgr_three_connection_mode {
 	PM_STA_SAP_SCC_24_SAP_5_DBS,
@@ -818,6 +835,12 @@ enum policy_mgr_three_connection_mode {
 	PM_NAN_DISC_NDI_24_NDI_5_DBS,
 	PM_NDI_NDI_5_NAN_DISC_24_DBS,
 	PM_NDI_NDI_NAN_DISC_24_SMM,
+	PM_SAP_SAP_SCC_24_STA_5_DBS,
+	PM_SAP_SAP_SCC_24_STA_24_SCC,
+	PM_SAP_SAP_SCC_24_SAP_5_DBS,
+	PM_SAP_SAP_SCC_5_STA_24_DBS,
+	PM_SAP_SAP_SCC_5_STA_5_SCC,
+	PM_SAP_SAP_SCC_5_SAP_24_DBS,
 
 	PM_MAX_THREE_CONNECTION_MODE
 };
diff --git a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_action.c b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_action.c
index 55a3d7c..9c3f92e 100644
--- a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_action.c
+++ b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_action.c
@@ -1,6 +1,6 @@
 /*
  * Copyright (c) 2012-2021 The Linux Foundation. All rights reserved.
- * Copyright (c) 2021 Qualcomm Innovation Center, Inc. All rights reserved.
+ * Copyright (c) 2021, 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -810,8 +810,10 @@ policy_mgr_get_next_action(struct wlan_objmgr_psoc *psoc,
 	uint32_t num_connections = 0;
 	enum policy_mgr_one_connection_mode second_index = 0;
 	enum policy_mgr_two_connection_mode third_index = 0;
+	enum policy_mgr_three_connection_mode fourth_index = 0;
 	policy_mgr_next_action_two_connection_table_type *second_conn_table;
 	policy_mgr_next_action_three_connection_table_type *third_conn_table;
+	policy_mgr_next_action_four_connection_table_type *fourth_conn_table;
 	enum policy_mgr_band band;
 	struct policy_mgr_psoc_priv_obj *pm_ctx;
 	enum QDF_OPMODE new_conn_mode = QDF_MAX_NO_OF_MODE;
@@ -872,6 +874,17 @@ policy_mgr_get_next_action(struct wlan_objmgr_psoc *psoc,
 			psoc, session_id, ch_freq, reason);
 		*next_action = (*third_conn_table)[third_index][band];
 		break;
+	case 3:
+		fourth_index =
+			policy_mgr_get_fourth_connection_pcl_table_index(psoc);
+		if (PM_MAX_THREE_CONNECTION_MODE == fourth_index) {
+			policy_mgr_err(
+			"couldn't find index for 4th connection next action table");
+			return QDF_STATUS_E_FAILURE;
+		}
+		fourth_conn_table = next_action_four_connection_table;
+		*next_action = (*fourth_conn_table)[fourth_index][band];
+		break;
 	default:
 		policy_mgr_err("unexpected num_connections value %d",
 			num_connections);
diff --git a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_core.c b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_core.c
index 3c3633c..89dbc14 100644
--- a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_core.c
+++ b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_core.c
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2021 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -2084,6 +2085,53 @@ QDF_STATUS policy_mgr_get_connection_channels(struct wlan_objmgr_psoc *psoc,
 			}
 		}
 		*len = num_channels;
+	}else if (POLICY_MGR_PCL_ORDER_24G_ONLY == order) {
+		while (PM_CONC_CONNECTION_LIST_VALID_INDEX(conn_index)) {
+			if (WLAN_REG_IS_24GHZ_CH_FREQ(
+				    pm_conc_connection_list[conn_index].freq)
+				&& (*index < weight_len)) {
+				ch_freq_list[num_channels++] =
+						pm_conc_connection_list[
+						conn_index++].freq;
+				pcl_weight[(*index)++] = weight1;
+			} else {
+				conn_index++;
+			}
+		}
+		*len = num_channels;
+	} else if (POLICY_MGR_PCL_ORDER_5G_ONLY == order) {
+		while (PM_CONC_CONNECTION_LIST_VALID_INDEX(conn_index)) {
+			if (skip_dfs_channel &&
+			    wlan_reg_is_dfs_for_freq(
+			    pm_ctx->pdev,
+			    pm_conc_connection_list[conn_index].freq)) {
+				conn_index++;
+			} else if (WLAN_REG_IS_5GHZ_CH_FREQ(
+				   pm_conc_connection_list[conn_index].freq) &&
+				   (*index < weight_len)) {
+				ch_freq_list[num_channels++] =
+							pm_conc_connection_list[
+							conn_index++].freq;
+				pcl_weight[(*index)++] = weight1;
+			} else {
+				conn_index++;
+			}
+		}
+		conn_index = 0;
+		while (add_6ghz &&
+		       PM_CONC_CONNECTION_LIST_VALID_INDEX(conn_index)) {
+			bool is_6ghz_ch = WLAN_REG_IS_6GHZ_CHAN_FREQ(
+				pm_conc_connection_list[conn_index].freq);
+			if (is_6ghz_ch && (*index < weight_len)) {
+				ch_freq_list[num_channels++] =
+						pm_conc_connection_list[
+						conn_index++].freq;
+				pcl_weight[(*index)++] = weight2;
+			} else {
+				conn_index++;
+			}
+		}
+		*len = num_channels;
 	} else {
 		policy_mgr_err("unknown order %d", order);
 		status = QDF_STATUS_E_FAILURE;
@@ -2471,6 +2519,28 @@ QDF_STATUS policy_mgr_get_channel_list(struct wlan_objmgr_psoc *psoc,
 		*len += num_channels;
 		status = QDF_STATUS_SUCCESS;
 		break;
+	case PM_SCC_ON_2G:
+		policy_mgr_get_connection_channels(
+			psoc, mode,
+			channel_list, &num_channels, POLICY_MGR_PCL_ORDER_24G_ONLY,
+			skip_dfs_channel, pcl_weights, weight_len, &i,
+			POLICY_MGR_PCL_GROUP_ID1_ID2);
+		qdf_mem_copy(&pcl_channels[*len], channel_list,
+			     num_channels * sizeof(*pcl_channels));
+		*len += num_channels;
+		status = QDF_STATUS_SUCCESS;
+		break;
+	case PM_SCC_ON_5G:
+		policy_mgr_get_connection_channels(
+			psoc, mode,
+			channel_list, &num_channels, POLICY_MGR_PCL_ORDER_5G_ONLY,
+			skip_dfs_channel, pcl_weights, weight_len, &i,
+			POLICY_MGR_PCL_GROUP_ID1_ID2);
+		qdf_mem_copy(&pcl_channels[*len], channel_list,
+			     num_channels * sizeof(*pcl_channels));
+		*len += num_channels;
+		status = QDF_STATUS_SUCCESS;
+		break;
 	case PM_SCC_ON_24_SCC_ON_5:
 		policy_mgr_get_connection_channels(
 			psoc, mode,
@@ -2684,6 +2754,26 @@ QDF_STATUS policy_mgr_get_channel_list(struct wlan_objmgr_psoc *psoc,
 		}
 		status = QDF_STATUS_SUCCESS;
 		break;
+	case PM_5G_24G:
+		policy_mgr_add_5g_to_pcl(
+			psoc, pcl_channels, len,
+			pcl_weights, weight_len,
+			&i,
+			POLICY_MGR_PCL_GROUP_ID1_ID2,
+			channel_list_5, chan_index_5,
+			channel_list_6, chan_index_6);
+
+		chan_index_24 = QDF_MIN((*len + chan_index_24),
+					weight_len) - *len;
+		if (chan_index_24 > 0) {
+			qdf_mem_copy(&pcl_channels[*len], channel_list_24,
+				     chan_index_24 * sizeof(*pcl_channels));
+			*len += chan_index_24;
+			for (j = 0; j < chan_index_24; i++, j++)
+				pcl_weights[i] = WEIGHT_OF_GROUP3_PCL_CHANNELS;
+		}
+		status = QDF_STATUS_SUCCESS;
+		break;
 	default:
 		policy_mgr_err("unknown pcl value %d", pcl);
 		break;
@@ -2745,35 +2835,98 @@ bool policy_mgr_disallow_mcc(struct wlan_objmgr_psoc *psoc,
 }
 
 /**
- * policy_mgr_is_multi_ap_plus_sta_3vif_conc() - Check multiple AP plus STA
- * concurrency
- * @mode1: policy_mgr_con_mode of connection 1
- * @mode2: policy_mgr_con_mode of connection 2
- * @mode2: policy_mgr_con_mode of connection 3
+ * policy_mgr_allow_same_mac_diff_freq() - Check whether diff freq are allowed
+ * on same mac
+ *
+ * @psoc: Pointer to Psoc
+ * @ch_freq: channel frequency
  *
- * Check the 3vif concurrency is SAP(GO)+SAP(GO)+STA or not based on
- * connection mode.
+ * Check whether diff freq are allowed on same mac
  *
  * Return: True/False
  */
-static bool policy_mgr_is_multi_ap_plus_sta_3vif_conc(
-	enum policy_mgr_con_mode mode1, enum policy_mgr_con_mode mode2,
-	enum policy_mgr_con_mode mode3)
+static
+bool policy_mgr_allow_same_mac_diff_freq(struct wlan_objmgr_psoc *psoc,
+					 qdf_freq_t ch_freq)
+{
+	bool allow = true;
+
+	if ((pm_conc_connection_list[0].mode == PM_NAN_DISC_MODE &&
+	     pm_conc_connection_list[1].mode == PM_NDI_MODE) ||
+	    (pm_conc_connection_list[0].mode == PM_NDI_MODE &&
+	     pm_conc_connection_list[1].mode == PM_NAN_DISC_MODE)) {
+		/*
+		 * NAN + NDI are managed in Firmware by dividing
+		 * up slots. Connection on NDI is re-negotiable
+		 * and therefore a 3rd connection with the
+		 * same MAC is possible.
+		 */
+	} else if (!policy_mgr_is_hw_dbs_capable(psoc) &&
+		   policy_mgr_is_interband_mcc_supported(psoc)) {
+		if (ch_freq !=  pm_conc_connection_list[0].freq &&
+		    ch_freq !=  pm_conc_connection_list[1].freq) {
+			policy_mgr_rl_debug("don't allow 3rd home channel on same MAC");
+			allow = false;
+		}
+	} else if (((WLAN_REG_IS_24GHZ_CH_FREQ(ch_freq)) &&
+		  (WLAN_REG_IS_24GHZ_CH_FREQ
+		  (pm_conc_connection_list[0].freq)) &&
+		  (WLAN_REG_IS_24GHZ_CH_FREQ
+		  (pm_conc_connection_list[1].freq))) ||
+		  ((WLAN_REG_IS_5GHZ_CH_FREQ(ch_freq)) &&
+		  (WLAN_REG_IS_5GHZ_CH_FREQ
+		  (pm_conc_connection_list[0].freq)) &&
+		  (WLAN_REG_IS_5GHZ_CH_FREQ
+		  (pm_conc_connection_list[1].freq)))) {
+		  policy_mgr_rl_debug("don't allow 3rd home channel on same MAC");
+		allow = false;
+	}
+
+	return allow;
+}
+
+/**
+ * policy_mgr_allow_same_mac_same_freq() - check whether given frequency is
+ * allowed for same mac
+ *
+ * @psoc: Pointer to Psoc
+ * @ch_freq: channel frequency
+ * @mode: Concurrency Mode
+ *
+ * check whether given frequency is allowed for same mac
+ *
+ * Return: True/False
+ */
+static
+bool policy_mgr_allow_same_mac_same_freq(struct wlan_objmgr_psoc *psoc,
+					 qdf_freq_t ch_freq,
+					 enum policy_mgr_con_mode mode)
 {
-	if (mode1 == PM_STA_MODE &&
-	    (mode2 == PM_SAP_MODE || mode2 == PM_P2P_GO_MODE) &&
-	    (mode3 == PM_SAP_MODE || mode3 == PM_P2P_GO_MODE))
-		return true;
-	if (mode2 == PM_STA_MODE &&
-	    (mode1 == PM_SAP_MODE || mode1 == PM_P2P_GO_MODE) &&
-	    (mode3 == PM_SAP_MODE || mode3 == PM_P2P_GO_MODE))
-		return true;
-	if (mode3 == PM_STA_MODE &&
-	    (mode1 == PM_SAP_MODE || mode1 == PM_P2P_GO_MODE) &&
-	    (mode2 == PM_SAP_MODE || mode2 == PM_P2P_GO_MODE))
-		return true;
-
-	return false;
+	bool allow = true;
+
+	if (!policy_mgr_is_hw_dbs_capable(psoc) &&
+	    policy_mgr_is_interband_mcc_supported(psoc)) {
+		policy_mgr_rl_debug("allow 2 intf SCC + new intf ch %d for legacy hw",
+				    ch_freq);
+	} else if ((pm_conc_connection_list[0].mode == PM_NAN_DISC_MODE &&
+		    pm_conc_connection_list[1].mode == PM_NDI_MODE) ||
+		    (pm_conc_connection_list[0].mode == PM_NDI_MODE &&
+		    pm_conc_connection_list[1].mode == PM_NAN_DISC_MODE)) {
+		/*
+		 * NAN + NDI are managed in Firmware by dividing
+		 * up slots. Connection on NDI is re-negotiable
+		 * and therefore a 3rd connection with the
+		 * same MAC is possible.
+		 */
+	} else if (wlan_reg_is_same_band_freqs(ch_freq,
+		   pm_conc_connection_list[0].freq) &&
+		   !policy_mgr_is_3rd_conn_on_same_band_allowed(
+		   psoc, mode, ch_freq)) {
+		policy_mgr_rl_debug("don't allow 3rd home channel on same MAC for sta+multi-AP");
+		allow = false;
+	}
+
+	return allow;
 }
 
 /**
@@ -2796,111 +2949,50 @@ bool policy_mgr_allow_new_home_channel(
 {
 	bool status = true;
 	struct policy_mgr_psoc_priv_obj *pm_ctx;
-	uint32_t mcc_to_scc_switch;
+	bool on_same_mac = false, force_switch_without_dis = false;
 
 	pm_ctx = policy_mgr_get_context(psoc);
 	if (!pm_ctx) {
 		policy_mgr_err("Invalid Context");
 		return false;
 	}
-	mcc_to_scc_switch =
-		policy_mgr_get_mcc_to_scc_switch_mode(psoc);
+
+	force_switch_without_dis =
+		policy_mgr_get_mcc_to_scc_switch_mode(psoc) ==
+		QDF_MCC_TO_SCC_SWITCH_FORCE_PREFERRED_WITHOUT_DISCONNECTION;
 
 	qdf_mutex_acquire(&pm_ctx->qdf_conc_list_lock);
-	if (num_connections == 2) {
-	/* No SCC or MCC combination is allowed with / on DFS channel */
-		if ((mcc_to_scc_switch ==
-		QDF_MCC_TO_SCC_SWITCH_FORCE_PREFERRED_WITHOUT_DISCONNECTION) &&
-		is_dfs_ch &&
-		((pm_conc_connection_list[0].ch_flagext &
-		  (IEEE80211_CHAN_DFS | IEEE80211_CHAN_DFS_CFREQ2)) ||
-		 (pm_conc_connection_list[1].ch_flagext &
-		  (IEEE80211_CHAN_DFS | IEEE80211_CHAN_DFS_CFREQ2)))) {
+	if (num_connections == 3) {
+		status = policy_mgr_allow_4th_new_freq(psoc,
+					pm_conc_connection_list[0].freq,
+					pm_conc_connection_list[1].freq,
+					pm_conc_connection_list[2].freq,
+					ch_freq);
+	} else if (num_connections == 2) {
+		/* No SCC or MCC combination is allowed with / on DFS channel */
+		on_same_mac = pm_conc_connection_list[0].mac ==
+			      pm_conc_connection_list[1].mac;
+		if (force_switch_without_dis && is_dfs_ch &&
+		   ((pm_conc_connection_list[0].ch_flagext &
+		   (IEEE80211_CHAN_DFS | IEEE80211_CHAN_DFS_CFREQ2)) ||
+		   (pm_conc_connection_list[1].ch_flagext &
+		   (IEEE80211_CHAN_DFS | IEEE80211_CHAN_DFS_CFREQ2)))) {
 			policy_mgr_rl_debug("Existing DFS connection, new 3-port DFS connection is not allowed");
 			status = false;
-
 		} else if (((pm_conc_connection_list[0].freq !=
-				pm_conc_connection_list[1].freq)
-		|| (mcc_to_scc_switch ==
-		QDF_MCC_TO_SCC_SWITCH_FORCE_PREFERRED_WITHOUT_DISCONNECTION)
-		) && (pm_conc_connection_list[0].mac ==
-			pm_conc_connection_list[1].mac)) {
-			if ((pm_conc_connection_list[0].mode ==
-							PM_NAN_DISC_MODE &&
-				    pm_conc_connection_list[1].mode ==
-								PM_NDI_MODE) ||
-				   (pm_conc_connection_list[0].mode ==
-								PM_NDI_MODE &&
-				    pm_conc_connection_list[1].mode ==
-							    PM_NAN_DISC_MODE)) {
-				/*
-				 * NAN + NDI are managed in Firmware by dividing
-				 * up slots. Connection on NDI is re-negotiable
-				 * and therefore a 3rd connection with the
-				 * same MAC is possible.
-				 */
-				status = true;
-			} else if (!policy_mgr_is_hw_dbs_capable(psoc) &&
-				   policy_mgr_is_interband_mcc_supported(psoc)) {
-				if (ch_freq !=
-				    pm_conc_connection_list[0].freq &&
-				    ch_freq !=
-				    pm_conc_connection_list[1].freq) {
-					policy_mgr_rl_debug("don't allow 3rd home channel on same MAC");
-					status = false;
-				}
-			} else if (((WLAN_REG_IS_24GHZ_CH_FREQ(ch_freq)) &&
-				   (WLAN_REG_IS_24GHZ_CH_FREQ
-				   (pm_conc_connection_list[0].freq)) &&
-				   (WLAN_REG_IS_24GHZ_CH_FREQ
-				   (pm_conc_connection_list[1].freq))) ||
-				   ((WLAN_REG_IS_5GHZ_CH_FREQ(ch_freq)) &&
-				   (WLAN_REG_IS_5GHZ_CH_FREQ
-				   (pm_conc_connection_list[0].freq)) &&
-				   (WLAN_REG_IS_5GHZ_CH_FREQ
-				   (pm_conc_connection_list[1].freq)))) {
-				policy_mgr_rl_debug("don't allow 3rd home channel on same MAC");
-				status = false;
-			}
-		} else if (pm_conc_connection_list[0].mac ==
-			   pm_conc_connection_list[1].mac) {
-			/* Existing two connections are SCC */
-			if (!policy_mgr_is_hw_dbs_capable(psoc) &&
-			    policy_mgr_is_interband_mcc_supported(psoc)) {
-				/* keep legacy chip "allow" as it is */
-				policy_mgr_rl_debug("allow 2 intf SCC + new intf ch %d for legacy hw",
-						    ch_freq);
-			} else if ((pm_conc_connection_list[0].mode ==
-							    PM_NAN_DISC_MODE &&
-				    pm_conc_connection_list[1].mode ==
-								PM_NDI_MODE) ||
-				   (pm_conc_connection_list[0].mode ==
-								PM_NDI_MODE &&
-				    pm_conc_connection_list[1].mode ==
-							    PM_NAN_DISC_MODE)) {
-				/*
-				 * NAN + NDI are managed in Firmware by dividing
-				 * up slots. Connection on NDI is re-negotiable
-				 * and therefore a 3rd connection with the
-				 * same MAC is possible.
-				 */
-			} else if (wlan_reg_is_same_band_freqs(ch_freq,
-					pm_conc_connection_list[0].freq) &&
-				   policy_mgr_is_multi_ap_plus_sta_3vif_conc(
-					pm_conc_connection_list[0].mode,
-					pm_conc_connection_list[1].mode,
-					mode)) {
-				policy_mgr_rl_debug("don't allow 3rd home channel on same MAC - sta existing");
-				status = false;
-			}
+			     pm_conc_connection_list[1].freq) ||
+			     force_switch_without_dis) && on_same_mac) {
+			status = policy_mgr_allow_same_mac_diff_freq(psoc,
+								     ch_freq);
+		} else if (on_same_mac) {
+			status = policy_mgr_allow_same_mac_same_freq(psoc,
+								     ch_freq,
+								     mode);
 		}
 	} else if ((num_connections == 1) &&
-		   (mcc_to_scc_switch ==
-		QDF_MCC_TO_SCC_SWITCH_FORCE_PREFERRED_WITHOUT_DISCONNECTION) &&
-		is_dfs_ch &&
-		(pm_conc_connection_list[0].ch_flagext &
-		 (IEEE80211_CHAN_DFS | IEEE80211_CHAN_DFS_CFREQ2))) {
-
+		   force_switch_without_dis && is_dfs_ch &&
+		   (pm_conc_connection_list[0].ch_flagext &
+		   (IEEE80211_CHAN_DFS | IEEE80211_CHAN_DFS_CFREQ2))) {
 		policy_mgr_rl_debug("Existing DFS connection, new 2-port DFS connection is not allowed");
 		status = false;
 	} else if ((num_connections == 1) &&
diff --git a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_get_set_utils.c b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_get_set_utils.c
index 995725c..444e17e 100644
--- a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_get_set_utils.c
+++ b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_get_set_utils.c
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2021 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -751,6 +752,67 @@ void policy_mgr_init_dbs_hw_mode(struct wlan_objmgr_psoc *psoc,
 		pm_ctx->num_dbs_hw_modes));
 }
 
+#ifdef FEATURE_FOURTH_CONNECTION
+/**
+ * policy_mgr_4_freq_always_on_same_mac() - Function to check if 4 freq can
+ * lead to 4 home freq on same mac
+ * @freq1: Frequency 1
+ * @freq2: Frequency 2
+ * @freq3: Frequency 3
+ * @freq4: Frequency 4
+ *
+ * Return:true if any 4 freq cause 4 home frequency on same mac
+ *
+ */
+static bool
+policy_mgr_4_freq_always_on_same_mac(     qdf_freq_t freq1, qdf_freq_t freq2,
+				     qdf_freq_t freq3, qdf_freq_t freq4)
+{
+	if(!((pm_conc_connection_list[0].mac ==
+	      pm_conc_connection_list[1].mac) &&
+	      (pm_conc_connection_list[1].mac ==
+	      pm_conc_connection_list[2].mac)))
+		return false;
+
+	if (((WLAN_REG_IS_24GHZ_CH_FREQ(freq1)) &&
+	    (WLAN_REG_IS_24GHZ_CH_FREQ(freq2)) &&
+	    (WLAN_REG_IS_24GHZ_CH_FREQ(freq3)) &&
+	    (WLAN_REG_IS_24GHZ_CH_FREQ(freq4))) ||
+	    ((WLAN_REG_IS_5GHZ_CH_FREQ(freq1)) &&
+	    (WLAN_REG_IS_5GHZ_CH_FREQ(freq2)) &&
+	    (WLAN_REG_IS_5GHZ_CH_FREQ(freq3)) &&
+	    (WLAN_REG_IS_5GHZ_CH_FREQ(freq4)))  ||
+	    ((WLAN_REG_IS_6GHZ_CHAN_FREQ(freq1)) &&
+	    (WLAN_REG_IS_6GHZ_CHAN_FREQ(freq2)) &&
+	    (WLAN_REG_IS_6GHZ_CHAN_FREQ(freq3)) &&
+	    (WLAN_REG_IS_6GHZ_CHAN_FREQ(freq4)))) {
+		policy_mgr_rl_debug("don't allow 4th home channel on same MAC");
+		return true;
+	}
+
+	return false;
+}
+
+bool
+policy_mgr_allow_4th_new_freq(struct wlan_objmgr_psoc *psoc,
+			      qdf_freq_t freq1, qdf_freq_t freq2,
+			      qdf_freq_t freq3, qdf_freq_t new_ch_freq)
+{
+	/* if HW is not DBS return false */
+	if (!policy_mgr_is_hw_dbs_capable(psoc))
+		return false;
+
+	if (!freq1 || !freq2 || !freq3 || !new_ch_freq) {
+		policy_mgr_info("one or more freq are 0: freq1 %d freq2 %d freq3 %d new_freq %d",
+				freq1, freq2, freq3, new_ch_freq);
+		return false;
+	}
+
+	return !policy_mgr_4_freq_always_on_same_mac(freq1, freq2, freq3,
+						     new_ch_freq);
+}
+#endif
+
 void policy_mgr_dump_dbs_hw_mode(struct wlan_objmgr_psoc *psoc)
 {
 	uint32_t i, param;
@@ -2653,7 +2715,8 @@ policy_mgr_allow_concurrency_csa(struct wlan_objmgr_psoc *psoc,
 	qdf_mutex_acquire(&pm_ctx->qdf_conc_list_lock);
 
 	if (forced && (reason == CSA_REASON_UNSAFE_CHANNEL ||
-		       reason == CSA_REASON_DCS))
+		       reason == CSA_REASON_DCS ||
+		       reason == CSA_REASON_CONCURRENT_STA_CHANGED_CHANNEL))
 		policy_mgr_store_and_del_conn_info_by_chan_and_mode(
 			psoc, old_ch_freq, mode, info, &num_cxn_del);
 	else
diff --git a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_i.h b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_i.h
index 3de5003..a09fb07 100644
--- a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_i.h
+++ b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_i.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2021 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -215,6 +216,8 @@ extern policy_mgr_next_action_two_connection_table_type
 		*next_action_two_connection_table;
 extern policy_mgr_next_action_three_connection_table_type
 		*next_action_three_connection_table;
+extern policy_mgr_next_action_four_connection_table_type
+		*next_action_four_connection_table;
 
 #ifdef FEATURE_FOURTH_CONNECTION
 extern const enum policy_mgr_pcl_type
diff --git a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_init_deinit.c b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_init_deinit.c
index 8a028ae..3e700c6 100644
--- a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_init_deinit.c
+++ b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_init_deinit.c
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2021 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -568,6 +569,8 @@ QDF_STATUS policy_mgr_psoc_enable(struct wlan_objmgr_psoc *psoc)
 		next_action_three_connection_table =
 		&pm_next_action_three_connection_dbs_1x1_table;
 	}
+	next_action_four_connection_table =
+		&pm_next_action_four_connection_dbs_2x2_table;
 	policy_mgr_debug("is DBS Capable %d, is SBS Capable %d",
 			 policy_mgr_is_hw_dbs_capable(psoc),
 			 policy_mgr_is_hw_sbs_capable(psoc));
diff --git a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_pcl.c b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_pcl.c
index b545802..ed0d28f 100644
--- a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_pcl.c
+++ b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_pcl.c
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2020 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -68,6 +69,9 @@ policy_mgr_next_action_two_connection_table_type
 		*next_action_two_connection_2x2_2g_1x1_5g_table;
 policy_mgr_next_action_three_connection_table_type
 		*next_action_three_connection_2x2_2g_1x1_5g_table;
+policy_mgr_next_action_four_connection_table_type
+		*next_action_four_connection_table;
+
 
 QDF_STATUS policy_mgr_get_pcl_for_existing_conn(
 		struct wlan_objmgr_psoc *psoc,
@@ -1945,7 +1949,33 @@ enum policy_mgr_three_connection_mode
 				psoc, PM_NAN_DISC_MODE, list_nan_disc);
 	policy_mgr_debug("sap/ago %d, sta %d, ndi %d nan disc %d",
 			 count_sap, count_sta, count_ndi, count_nan_disc);
-	if (count_sap == 2 && count_sta == 1) {
+	if (count_sap == 3) {
+		if (WLAN_REG_IS_SAME_BAND_FREQS(
+			pm_conc_connection_list[list_sap[0]].freq,
+			pm_conc_connection_list[list_sap[1]].freq)) {
+			if (WLAN_REG_IS_24GHZ_CH_FREQ(
+				pm_conc_connection_list[list_sap[2]].freq))
+				index = PM_SAP_SAP_SCC_5_SAP_24_DBS;
+			else
+				index = PM_SAP_SAP_SCC_24_SAP_5_DBS;
+		} else if (WLAN_REG_IS_SAME_BAND_FREQS(
+			pm_conc_connection_list[list_sap[0]].freq,
+			pm_conc_connection_list[list_sap[2]].freq)) {
+			if (WLAN_REG_IS_24GHZ_CH_FREQ(
+				pm_conc_connection_list[list_sap[1]].freq))
+				index = PM_SAP_SAP_SCC_5_SAP_24_DBS;
+			else
+				index = PM_SAP_SAP_SCC_24_SAP_5_DBS;
+		} else if (WLAN_REG_IS_SAME_BAND_FREQS(
+			pm_conc_connection_list[list_sap[1]].freq,
+			pm_conc_connection_list[list_sap[2]].freq)) {
+			if (WLAN_REG_IS_24GHZ_CH_FREQ(
+				pm_conc_connection_list[list_sap[0]].freq))
+				index = PM_SAP_SAP_SCC_5_SAP_24_DBS;
+			else
+				index = PM_SAP_SAP_SCC_24_SAP_5_DBS;
+		}
+	} else if (count_sap == 2 && count_sta == 1) {
 		policy_mgr_debug(
 			"channel: sap0: %d, sap1: %d, sta0: %d",
 			pm_conc_connection_list[list_sap[0]].freq,
@@ -1979,6 +2009,34 @@ enum policy_mgr_three_connection_mode
 		     WLAN_REG_IS_5GHZ_CH_FREQ(
 			pm_conc_connection_list[list_sap[0]].freq)) {
 			index = PM_STA_SAP_SCC_5_SAP_24_DBS;
+		} else if (WLAN_REG_IS_24GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sap[1]].freq) &&
+		     WLAN_REG_IS_5GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sta[0]].freq) &&
+		     WLAN_REG_IS_24GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sap[0]].freq)) {
+			index = PM_SAP_SAP_SCC_24_STA_5_DBS;
+		} else if (WLAN_REG_IS_5GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sap[1]].freq) &&
+		     WLAN_REG_IS_24GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sta[0]].freq) &&
+		     WLAN_REG_IS_5GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sap[0]].freq)) {
+			index = PM_SAP_SAP_SCC_5_STA_24_DBS;
+		} else if (WLAN_REG_IS_24GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sap[1]].freq) &&
+		     WLAN_REG_IS_24GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sap[0]].freq) &&
+		     WLAN_REG_IS_24GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sta[0]].freq)) {
+			index = PM_SAP_SAP_SCC_24_STA_24_SCC;
+		} else if (WLAN_REG_IS_5GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sap[1]].freq) &&
+		     WLAN_REG_IS_5GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sap[0]].freq) &&
+		     WLAN_REG_IS_5GHZ_CH_FREQ(
+			pm_conc_connection_list[list_sta[0]].freq)) {
+			index = PM_SAP_SAP_SCC_5_STA_5_SCC;
 		} else {
 			index =  PM_MAX_THREE_CONNECTION_MODE;
 		}
@@ -2654,3 +2712,96 @@ QDF_STATUS policy_mgr_filter_passive_ch(struct wlan_objmgr_pdev *pdev,
 
 	return QDF_STATUS_SUCCESS;
 }
+
+bool policy_mgr_is_3rd_conn_on_same_band_allowed(struct wlan_objmgr_psoc *psoc,
+						 enum policy_mgr_con_mode mode,
+						 qdf_freq_t ch_freq)
+{
+	enum policy_mgr_pcl_type pcl = PM_NONE;
+	enum policy_mgr_conc_priority_mode conc_system_pref = 0;
+	enum policy_mgr_two_connection_mode third_index = 0;
+	struct policy_mgr_psoc_priv_obj *pm_ctx;
+	bool ret = false;
+
+	pm_ctx = policy_mgr_get_context(psoc);
+	if (!pm_ctx) {
+		policy_mgr_err("context is NULL");
+			return false;
+	}
+
+	if (pm_conc_connection_list[0].freq != ch_freq ||
+	    pm_conc_connection_list[0].freq !=
+				pm_conc_connection_list[1].freq) {
+		policy_mgr_debug("No MCC support in 3vif in same mac: %d %d %d",
+				 pm_conc_connection_list[0].freq,
+				 pm_conc_connection_list[1].freq,
+				 ch_freq);
+		return false;
+	}
+
+	policy_mgr_debug("pref:%d requested mode:%d",
+			 pm_ctx->cur_conc_system_pref, mode);
+
+	switch (pm_ctx->cur_conc_system_pref) {
+	case 0:
+		conc_system_pref = PM_THROUGHPUT;
+		break;
+	case 1:
+		conc_system_pref = PM_POWERSAVE;
+		break;
+	case 2:
+		conc_system_pref = PM_LATENCY;
+		break;
+	default:
+		policy_mgr_err("unknown cur_conc_system_pref value %d",
+			       pm_ctx->cur_conc_system_pref);
+		break;
+	}
+
+	third_index = policy_mgr_get_third_connection_pcl_table_index(psoc);
+	policy_mgr_debug("pcl for third connection is %d", third_index);
+	if (PM_MAX_TWO_CONNECTION_MODE == third_index) {
+		policy_mgr_err(
+			"couldn't find index for 3rd connection pcl table");
+			return false;
+	}
+	if (policy_mgr_is_hw_dbs_capable(psoc) == true) {
+		pcl = (*third_connection_pcl_dbs_table)
+			[third_index][mode][conc_system_pref];
+	} else {
+		pcl = (*third_connection_pcl_non_dbs_table)
+			[third_index][mode][conc_system_pref];
+	}
+
+	policy_mgr_debug("pcl for third connection mode %s is %d %s",
+			 device_mode_to_string(mode), pcl,
+			 pcl_type_to_string(pcl));
+	switch (pcl) {
+	case PM_SCC_CH:
+	case PM_SCC_CH_24G:
+	case PM_SCC_CH_5G:
+	case PM_24G_SCC_CH:
+	case PM_5G_SCC_CH:
+	case PM_SCC_ON_5_SCC_ON_24_24G:
+	case PM_SCC_ON_5_SCC_ON_24_5G:
+	case PM_SCC_ON_24_SCC_ON_5_24G:
+	case PM_SCC_ON_24_SCC_ON_5_5G:
+	case PM_SCC_ON_5_SCC_ON_24:
+	case PM_SCC_ON_24_SCC_ON_5:
+	case PM_24G_SCC_CH_SBS_CH:
+	case PM_24G_SCC_CH_SBS_CH_5G:
+	case PM_MCC_CH:
+	case PM_MCC_CH_24G:
+	case PM_MCC_CH_5G:
+	case PM_24G_MCC_CH:
+	case PM_5G_MCC_CH:
+	case PM_24G_SBS_CH_MCC_CH:
+		ret = true;
+		break;
+	default:
+		policy_mgr_debug("Not in SCC case");
+		ret = false;
+		break;
+	}
+	return ret;
+}
diff --git a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_tables_2x2_dbs_i.h b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_tables_2x2_dbs_i.h
index 4d1d964..340fb03 100644
--- a/components/cmn_services/policy_mgr/src/wlan_policy_mgr_tables_2x2_dbs_i.h
+++ b/components/cmn_services/policy_mgr/src/wlan_policy_mgr_tables_2x2_dbs_i.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2012-2020 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -238,7 +239,7 @@ pm_third_connection_pcl_dbs_2x2_table = {
 
 	[PM_STA_SAP_SCC_24_2x2] = {
 	[PM_STA_MODE] = {PM_5G_SCC_CH, PM_5G_SCC_CH, PM_5G_SCC_CH},
-	[PM_SAP_MODE] = {PM_5G, PM_5G, PM_5G},
+	[PM_SAP_MODE] = {PM_5G_SCC_CH, PM_5G_SCC_CH, PM_5G_SCC_CH},
 	[PM_P2P_CLIENT_MODE] = {PM_5G, PM_5G, PM_5G},
 	[PM_P2P_GO_MODE] = {PM_5G, PM_5G, PM_5G},
 	[PM_NAN_DISC_MODE] = {
@@ -276,7 +277,7 @@ pm_third_connection_pcl_dbs_2x2_table = {
 	[PM_STA_SAP_SCC_5_2x2] = {
 	[PM_STA_MODE] = {PM_24G_SCC_CH_SBS_CH_5G, PM_24G_SCC_CH,
 		PM_24G_SCC_CH_SBS_CH},
-	[PM_SAP_MODE] = {PM_24G, PM_24G, PM_24G},
+	[PM_SAP_MODE] = {PM_24G_SCC_CH, PM_24G_SCC_CH, PM_24G_SCC_CH},
 	[PM_P2P_CLIENT_MODE] = {PM_24G, PM_24G, PM_24G},
 	[PM_P2P_GO_MODE] = {PM_24G, PM_24G, PM_24G},
 	[PM_NAN_DISC_MODE] = {
@@ -1786,6 +1787,10 @@ const enum policy_mgr_pcl_type
 fourth_connection_pcl_dbs_table
 	[PM_MAX_THREE_CONNECTION_MODE][PM_MAX_NUM_OF_MODE]
 	[PM_MAX_CONC_PRIORITY_MODE] = {
+	[PM_STA_SAP_SCC_24_SAP_5_DBS] = {
+	[PM_SAP_MODE] = { PM_SCC_ON_5_SCC_ON_24, PM_SCC_ON_5_SCC_ON_24, PM_SCC_ON_5_SCC_ON_24 } },
+	[PM_STA_SAP_SCC_5_SAP_24_DBS] = {
+	[PM_SAP_MODE] = { PM_SCC_ON_24_SCC_ON_5, PM_SCC_ON_24_SCC_ON_5, PM_SCC_ON_24_SCC_ON_5 } },
 	[PM_NAN_DISC_STA_24_NDI_5_DBS] = {
 	[PM_NDI_MODE] = { PM_5G, PM_5G, PM_5G } },
 	[PM_NAN_DISC_NDI_24_STA_5_DBS] = {
@@ -1799,7 +1804,19 @@ fourth_connection_pcl_dbs_table
 	[PM_NDI_NDI_5_NAN_DISC_24_DBS] = {
 	[PM_STA_MODE] = { PM_5G, PM_5G, PM_5G } },
 	[PM_NDI_NDI_NAN_DISC_24_SMM] = {
-	[PM_STA_MODE] = { PM_5G, PM_5G, PM_5G } }
+	[PM_STA_MODE] = { PM_5G, PM_5G, PM_5G } },
+	[PM_SAP_SAP_SCC_24_SAP_5_DBS] = {
+	[PM_STA_MODE] = { PM_5G_24G, PM_5G_24G, PM_5G_24G } },
+	[PM_SAP_SAP_SCC_24_STA_5_DBS] = {
+	[PM_SAP_MODE] = { PM_SCC_ON_5G, PM_SCC_ON_5G, PM_SCC_ON_5G } },
+	[PM_SAP_SAP_SCC_5_SAP_24_DBS] = {
+	[PM_STA_MODE] = { PM_5G_24G, PM_5G_24G, PM_5G_24G } },
+	[PM_SAP_SAP_SCC_5_STA_24_DBS] = {
+	[PM_SAP_MODE] = { PM_SCC_ON_2G, PM_SCC_ON_2G, PM_SCC_ON_2G } },
+	[PM_SAP_SAP_SCC_24_STA_24_SCC] = {
+	[PM_SAP_MODE] = { PM_5G, PM_5G, PM_5G } },
+	[PM_SAP_SAP_SCC_5_STA_5_SCC] = {
+	[PM_SAP_MODE] = { PM_24G, PM_24G, PM_24G } },
 };
 #endif
 #endif
@@ -1938,4 +1955,18 @@ static policy_mgr_next_action_three_connection_table_type
 	[PM_P2P_GO_SAP_SBS_5_1x1] = {PM_DBS_UPGRADE, PM_NOP},
 };
 
+/**
+ * next_action_three_connection_table - table which provides next
+ * action while a new connection is coming up, with two
+ * connections already in the system
+ */
+static policy_mgr_next_action_four_connection_table_type
+	pm_next_action_four_connection_dbs_2x2_table = {
+	[PM_SAP_SAP_SCC_24_STA_5_DBS] = {PM_NOP,	PM_NOP},
+	[PM_SAP_SAP_SCC_24_STA_24_SCC] = {PM_DBS,	PM_DBS},
+	[PM_SAP_SAP_SCC_24_SAP_5_DBS] = {PM_NOP,	PM_NOP},
+	[PM_SAP_SAP_SCC_5_STA_24_DBS] = {PM_NOP,	PM_NOP},
+	[PM_SAP_SAP_SCC_5_STA_5_SCC] = {PM_DBS,	PM_DBS},
+	[PM_SAP_SAP_SCC_5_SAP_24_DBS] = {PM_NOP,	PM_NOP},
+};
 #endif
diff --git a/core/hdd/src/wlan_hdd_conc_ut.c b/core/hdd/src/wlan_hdd_conc_ut.c
index 1594e84..eb8f326 100644
--- a/core/hdd/src/wlan_hdd_conc_ut.c
+++ b/core/hdd/src/wlan_hdd_conc_ut.c
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2015-2020 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023, Qualcomm Innovation Center, Inc. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for
  * any purpose with or without fee is hereby granted, provided that the
@@ -96,45 +97,6 @@ static const char *system_config_to_string(uint8_t idx)
 
 }
 
-static const char *device_mode_to_string(uint8_t idx)
-{
-	switch (idx) {
-	CASE_RETURN_STRING(PM_STA_MODE);
-	CASE_RETURN_STRING(PM_SAP_MODE);
-	CASE_RETURN_STRING(PM_P2P_CLIENT_MODE);
-	CASE_RETURN_STRING(PM_P2P_GO_MODE);
-	default:
-		return "none";
-	}
-}
-
-static const char *pcl_type_to_string(uint8_t idx)
-{
-	switch (idx) {
-	CASE_RETURN_STRING(PM_NONE);
-	CASE_RETURN_STRING(PM_24G);
-	CASE_RETURN_STRING(PM_5G);
-	CASE_RETURN_STRING(PM_SCC_CH);
-	CASE_RETURN_STRING(PM_MCC_CH);
-	CASE_RETURN_STRING(PM_SCC_CH_24G);
-	CASE_RETURN_STRING(PM_SCC_CH_5G);
-	CASE_RETURN_STRING(PM_24G_SCC_CH);
-	CASE_RETURN_STRING(PM_5G_SCC_CH);
-	CASE_RETURN_STRING(PM_SCC_ON_5_SCC_ON_24_24G);
-	CASE_RETURN_STRING(PM_SCC_ON_5_SCC_ON_24_5G);
-	CASE_RETURN_STRING(PM_SCC_ON_24_SCC_ON_5_24G);
-	CASE_RETURN_STRING(PM_SCC_ON_24_SCC_ON_5_5G);
-	CASE_RETURN_STRING(PM_SCC_ON_5_SCC_ON_24);
-	CASE_RETURN_STRING(PM_SCC_ON_24_SCC_ON_5);
-	CASE_RETURN_STRING(PM_MCC_CH_24G);
-	CASE_RETURN_STRING(PM_MCC_CH_5G);
-	CASE_RETURN_STRING(PM_24G_MCC_CH);
-	CASE_RETURN_STRING(PM_5G_MCC_CH);
-	default:
-		return "Unknown";
-	}
-}
-
 void clean_report(struct hdd_context *hdd_ctx)
 {
 	uint32_t idx = 0;
diff --git a/core/hdd/src/wlan_hdd_hostapd.c b/core/hdd/src/wlan_hdd_hostapd.c
index fc04613..ce1c13c 100644
--- a/core/hdd/src/wlan_hdd_hostapd.c
+++ b/core/hdd/src/wlan_hdd_hostapd.c
@@ -3272,6 +3272,7 @@ QDF_STATUS wlan_hdd_get_channel_for_sap_restart(
 				struct wlan_objmgr_psoc *psoc,
 				uint8_t vdev_id, uint32_t *ch_freq)
 {
+	bool is_force_scc;
 	mac_handle_t mac_handle;
 	struct hdd_ap_ctx *hdd_ap_ctx;
 	struct hdd_context *hdd_ctx;
@@ -3402,9 +3403,10 @@ sap_restart:
 		  ch_params.ch_width);
 	hdd_ap_ctx->bss_stop_reason = BSS_STOP_DUE_TO_MCC_SCC_SWITCH;
 	*ch_freq = intf_ch_freq;
+	is_force_scc = policy_mgr_is_force_scc(psoc);
 	hdd_debug("SAP channel change with CSA/ECSA");
 	hdd_sap_restart_chan_switch_cb(psoc, vdev_id, *ch_freq,
-				       ch_params.ch_width, false);
+				       ch_params.ch_width, is_force_scc);
 	wlansap_context_put(sap_context);
 
 	return QDF_STATUS_SUCCESS;
-- 
2.7.4

